Vector store demo

lazeratops opened the pull request:
This PR contains a vector store and querying demo for Daily recordings or manually uploaded video files. Please see the README for more info.

lazeratops committed:
Very very wip

lazeratops committed:
Manual upload flow wip

lazeratops committed:
Add tip

lazeratops committed:
res.response is not an array.

lazeratops committed:
Merge main

lazeratops committed:
Commenting server-side

lazeratops committed:
Cleanup, recording vs upload storage separation

lazeratops committed:
autopep8

lazeratops committed:
Aggressive autopep8

lazeratops committed:
pylint

lazeratops committed:
eslint

lazeratops committed:
some error handling tweaks

lazeratops committed:
Styling, cleanup, persistence

lazeratops committed:
README update

lazeratops commented on server/store.py:
```diff
@@ -0,0 +1,337 @@
+        # Wait for tasks to complete
+        await asyncio.gather(*tasks)
+
+    async def index_daily_recording(self, sem: asyncio.locks.BoundedSemaphore,
```
I think this is wonky but I'm keeping it here atm because it technically "works" and can be a starting point for feedback. I am using a bounded semaphore here to limit us to processing 5 recordings at a time, but I have a feeling it may be better to create a single reusable executor instance on this class with a `max_workers` parameter instead. WDYT? I think the task -> executor nesting is also potentially not needed.

aconchillo replied:
this looks neat! but yes probably an executor would be simpler?

```
with ThreadPoolExecutor(max_workers=5) as executor:
        tasks = []
        for recording in recordings:
            tasks.append(asyncio.create_task(self.index_daily_recording(executor, recording)))
        # Wait for tasks to complete
        await asyncio.gather(*tasks)
```

and then:

```
async def index_daily_recording(self, executor: asyncio.concurrent.futures.Executor, recording: Recording):
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(executor, self.transcribe_and_index_recording, recording)
```

I think you need the tasks otherwise you will just be creating one thread (and waiting for it with `await loop.run_in_executor(...)`) at a time instead of 5 and you won't know when they are all finished.

lazeratops replied:
I think what can make the task redundant is that I don't need to await each executor run. I could invoke `async.run_in_executor()` as many times as I need and then `await async.gather()` them all at the end. I've just pushed the approach I'm proposing there. This should:

* Remove added layer of "task" creation (I still call the futures returned from `run_in_executor()` "tasks" which I then gather, but get rid of `create_task()`.
* Enable removal an extra "nested" level of methods in the store.
* Allow for cancellation of futures if server shutdown is invoked.

Does this look reasonable?

aconchillo replied:
Ah, yes! Perfect!

aconchillo replied:
I didn't realize `run_in_executor` returned a future. :raised_hands: !

lazeratops committed:
Check whether file is still being written better

lazeratops commented on client/index.js:
```diff
@@ -0,0 +1,245 @@
+
+window.addEventListener('DOMContentLoaded', () => {
+  fetchCapabilities().then((capabilities) => {
+    if (capabilities.daily === true) {
```
This will be false if the Daily API key is not specified in the server environment.

lazeratops committed:
Cleanup

lazeratops commented:
@jessmitch42  Would love your input on the frontend bits here!

lazeratops committed:
Cleanup

lazeratops committed:
Tiny cleanup

lazeratops committed:
Tiny cleanup

aconchillo commented on server/store.py:
```diff
@@ -0,0 +1,337 @@
+            tasks.append(asyncio.create_task(
+                self.index_uploaded_file(sem, path)))
+        # Wait for all processing tasks to finish
+        await asyncio.gather(*tasks)
```
same executor approach here.

aconchillo commented on server/store.py:
```diff
@@ -0,0 +1,337 @@
+        transcript_file_name = f"{file_name}.txt"
+        transcripts_dir = get_transcripts_dir_path()
+        transcript_file_path = os.path.join(
+            transcripts_dir, transcript_file_name)
```
may be a function to get the transcript file name since we use it in `transcribe_and_index_file` as well. just in case we change in one place but not the other.

aconchillo replied:
```
def get_transcript_file_path(file_name):
        transcript_file_name = f"{file_name}.txt"
        transcripts_dir = get_transcripts_dir_path()
        return os.path.join(transcripts_dir, transcript_file_name)
```

lazeratops replied:
Good idea

aconchillo commented on server/store.py:
```diff
@@ -0,0 +1,337 @@
+            if source == Source.DAILY:
+                await self.index_daily_recordings()
+            elif source == Source.UPLOADS:
+                await self.generate_upload_transcripts()
```
maybe this could be:

```
await self.generate_index(source, False)
```

where `False` is from:

```
    async def generate_index(self, source: Source, create_index: bool):
        """Generates a new index from given source"""
        if source == Source.DAILY:
            await self.index_daily_recordings()
        elif source == Source.UPLOADS:
            await self.generate_upload_transcripts()
        if (create_index):
            self.create_index()
```

lazeratops replied:
🤦  Of course

aconchillo commented on server/store.py:
```diff
@@ -0,0 +1,337 @@
+        if not self.ready():
+            self.update_status(State.CREATING, "Creating index")
+            try:
+                await self.generate_index(source)
```
and here:

```
                await self.generate_index(source, True)
```

aconchillo commented:
This looks awesome! Just added a few comments.

Also, instead of loading `os.getenv()` in different functions, I would suggest creating a `Config` class with all the values you need and pass that instead in the constructors (e.g. Store and Transcriber). This way functions are more testable.

jessmitch42 commented on client/css/style.css:
```diff
@@ -0,0 +1,235 @@
+}
+
+.disabled {
+    background-color: var(--dark-muted);
```
Nit: I would make this a bit lighter (the background color) -- either #737374 (or whatever you colour var equivalent is) or opacity:0.6. And then add a cursor to indicate it can't be clicked
```suggestion
    background-color: var(--dark-muted);
    cursor: not-allowed;
```

jessmitch42 commented on client/css/style.css:
```diff
@@ -0,0 +1,235 @@
+
+form {
+    margin-bottom: 8px;
+    margin-bottom: 8px;
```
```suggestion
```
repeated

jessmitch42 commented on client/css/style.css:
```diff
@@ -0,0 +1,235 @@
+
+form button {
+    font-size: 1.2em;
+    margin: 4px;
```
```suggestion
    margin: 8px 0;
```
Nit: not every button has a horizontal margin so I would just remove it for consistency.

jessmitch42 commented on client/css/style.css:
```diff
@@ -0,0 +1,235 @@
+
+.dark-btn {
+    background-color: var(--dark);
+    border-color: var(--dark);
```
Little note: usually `border: colour line-style width` is easier to read so you could put `border: var(--dark) solid 1px;`
Doesn't matter, just fewer lines and easier to skim as a reader

lazeratops replied:
Ty, done!

jessmitch42 reviewed (approved):
The frontend code looks great to me, nice work! No notes other than some tiny CSS suggestions

lazeratops committed:
Feedback + autopep8

lazeratops committed:
Removec some redundancies in the concurrency logic, use max_workers in a thread pool executor vs a bounded semaphore

lazeratops committed:
Update client/css/style.css

Co-authored-by: Jess Mitchell <jess@jessmitchell.ca>

lazeratops committed:
Update client/css/style.css

Co-authored-by: Jess Mitchell <jess@jessmitchell.ca>

lazeratops committed:
Update client/css/style.css

Co-authored-by: Jess Mitchell <jess@jessmitchell.ca>

lazeratops committed:
Update border style definition for .dark-btn

lazeratops committed:
Use config object and only retrieve env variables from `config` module.

lazeratops commented:
> This looks awesome! Just added a few comments.
> 
> Also, instead of loading `os.getenv()` in different functions, I would suggest creating a `Config` class with all the values you need and pass that instead in the constructors (e.g. Store and Transcriber). This way functions are more testable.

Thank you for the review! I think I've addressed everything now, including creating an "APIConfig" class for third-party keys, model-names, etc.

aconchillo commented on server/config.py:
```diff
@@ -0,0 +1,73 @@
+    deepgram_model_name: str = None
+
+
+def get_third_party_config() -> APIConfig:
```
Nitpick: Maybe this could be called `get_third_party_config_from_env()` or simply `from_env()` so it's clear we will be using environment variables.

aconchillo commented on server/config.py:
```diff
@@ -0,0 +1,73 @@
+
+def get_transcripts_dir_path() -> str:
+    """Returns transcript directory path."""
+    return os.path.abspath(os.getenv(TRANSCRIPTS_DIR_ENV))
```
I believe all these should also be part of the `APIConfig`. Because, again, I would not be able to build a valid `APIConfig` without environment variables.

So, this could be:

```
def get_transcripts_dir_path() -> str:
    """Returns transcript directory path."""
    return os.path.abspath(self.transcripts_dir)
```

lazeratops replied:
These are not part of any third-party API though (which was the thinking behind the name `APIConfig`), and the only one that consumes them is this specific file. I think including them in what essentially turns into a global object being explicitly passed around for consumption errs on exposing the side of too many variables to things that don't need them (additionally introducing a risk of accidental manipulation). Like, if the local directory path env variables are only ever needed in this one file, I sort of prefer not to expose them and pass them around further.

lazeratops replied:
Actually... I think you're right. I will rename this class, make a single Config class out of this file (more appropriate naming), and possibly make use of a read-only property approach or just the underscore prefix convention.

aconchillo replied:
Yeah, doesn't need to be only API keys, could be anything the application needs as configuration. :raised_hands:

lazeratops replied:
Alright, I submitted a new config class. Overview:

* The primary consumers of the `Config` instance are `index.py` server routes and the vector store. All other modules receive the variables they need directly vs the entire `Config` object (for isolation purposes and because they just don't need to know).
* All env variables are now retrieved only once, when initializing the config class, if no other values for each property are provided to the constructor.
* Default fallbacks remain the same for things like local directories.
* Using underscore convention for intended-private config variables and property accessors to retrieve them.

Hope that makes sense!

aconchillo replied:
Nice, thanks!

aconchillo commented on server/config.py:
```diff
@@ -0,0 +1,73 @@
+        os.getenv("DAILY_API_URL"),
+        os.getenv("DEEPGRAM_API_KEY"),
+        os.getenv("DEEPGRAM_MODEL_NAME")
+    )
```
Based one the other comment, this would be:

```
    return APIConfig(
        os.getenv("DAILY_API_KEY"),
        os.getenv("DAILY_API_URL"),
        os.getenv("DEEPGRAM_API_KEY"),
        os.getenv("DEEPGRAM_MODEL_NAME"),
        os.getenv("TRANSCRIPTS_DIR"),
        os.getenv("INDEX_DIR"),
        os.getenv("UPLOAD_DIR"),
        os.getenv("RECORDINGS_DIR"),
    )
```

aconchillo replied:
This would allow you to have the config in a JSON file for example:

```
def from_config_file(file_name):
   ....
```

aconchillo commented on server/config.py:
```diff
@@ -0,0 +1,73 @@
+def ensure_dir(env_name: str):
+    """Creates directory based on env variable,
+    if said directory does not already exist."""
+    directory = os.getenv(env_name)
```
this should just be:

```
def ensure_dir(directory: str):
    if not os.path.exists(directory):
        os.makedirs(directory)
```

aconchillo commented on server/config.py:
```diff
@@ -0,0 +1,73 @@
+    ensure_dir(TRANSCRIPTS_DIR_ENV)
+    ensure_dir(INDEX_DIR_ENV)
+    ensure_dir(UPLOAD_DIR_ENV)
+    ensure_dir(RECORDINGS_DIR_ENV)
```
This would be:

```
def ensure_dirs():
    """Creates required file directories if they do not already exist."""
    ensure_dir(self.transcripts_dir)
    ensure_dir(self.index_dir)
    ensure_dir(self.upload_dir)
    ensure_dir(self.recordings_dir)
```

aconchillo commented:
Approving! Just a few last comments about the new `APIConfig`. Hopefully they make sense. This is awesome!!!

lazeratops committed:
Create Config class

lazeratops commented on server/config.py:
```diff
@@ -0,0 +1,111 @@
+        """Returns transcript directory path."""
+        return self._recordings_dir_path
+
+    def get_transcript_file_path(self, file_name: str) -> str:
```
Note: Did not make this or `get_remote_recording_audio_path()` below a property because I'd rather have an explicit method call for anything that requires any computation other than retrieval of a class member.

aconchillo replied:
Yes, perfect!

lazeratops committed:
autopep8

aconchillo commented:
:ship: it!
